Add official support for AISE to the Python 3 version of PyMS

poiuyqwert opened the pull request.

poiuyqwert committed:
NOTE: This commit will not work without the next commit, just wanted to separate specific AISE stuff to a separate commit
NOTE: IScriptBIN has not been updated for new refactors
- Refactor CodeHandlers for work on AISE support
- Added concept of "LanguageDefinition" which takes "LanguagePlugins" to add extensions like AISE
- Split out a CompileContext and DecompileContext to closer match Parse/Serialize and add language context
- Further split CodeType's lex/parse for some easier subclassing
- Renamed Handler/Builder as Decompiler/Compiler

poiuyqwert committed:
- Added AISEContext to maintain AISE support state (like if the file is expanded, or the long jump mapping)
- Added AILanguage with core language and AISE plugin
- Started adding AISE commands
- Started creating AISE types and test for the types

poiuyqwert committed:
- Simplify CodeType lex/parse
- Add some more AISE types
- Fix/update some tests

poiuyqwert committed:
- Implement decompile of IdleOrderFlags

poiuyqwert committed:
- Made Option protocol for all idle order options (including OptionSet)
- Made OptionSet a concrete type
- Encapsulate OptionSet decompile into itself
- Implemented Option serialization
- Added OptionSet simplify

poiuyqwert committed:
- Add parsing

poiuyqwert committed:
- Add compiling for idle_orders

poiuyqwert committed:
- Complete AISE commands

poiuyqwert committed:
- Fix string terminator handling
- Allow UnitGroup to accept other unit based types
- Set wait_rand to separate code

poiuyqwert committed:
- Finalize loading and saving expanded AISE files
- Update CodeType generics to represent binary type and memory type
- Encapsulated CodeCommand param iterating into generator
- Warn users about loading AISE files, or when a change needs to activate AISE support

poiuyqwert committed:
- Update IScriptBIN to use new code handlers language setup
- Rename _lex back to lex

poiuyqwert committed:
- Update versions

poiuyqwert committed:
- Add HasKeywords to AISECodeTypes that were missing it

poiuyqwert committed:
- Update syntax highlighting for AISE

poiuyqwert committed:
- Fix PyICE code editor parsing preview parameters

poiuyqwert committed:
- Fix an issue with atomic saving on windows

poiuyqwert committed:
- Fix issue saving bwscripts

poiuyqwert committed:
- Working on load issue fixer

poiuyqwert committed:
- Refresh list on resolution updates

poiuyqwert committed:
- Fix EditScriptDialog
- Cleanup StringEditor code

poiuyqwert committed:
Merge branch 'python3' into AISE

# Conflicts:
#	PyMS/FileFormats/AIBIN/AIBIN.py
#	PyMS/versions.json

poiuyqwert committed:
- Switch pylint settings from "only manually enabled" to "only manually disabled" to try and catch more issues
- Fixed up PyAI pylint reports

poiuyqwert committed:
- Fixed up PyBIN lint reports

poiuyqwert committed:
- Fixed up PyDAT lint reports

poiuyqwert committed:
- Fixed up PyFNT lint reports

poiuyqwert committed:
- Forgot DAT lint updates

poiuyqwert committed:
- Forgot FNT lint updates

poiuyqwert committed:
Fix versions.json

poiuyqwert committed:
- Fixed up PyGOT lint reports

poiuyqwert committed:
- Fixed up PyGRP lint reports

poiuyqwert committed:
- Fixed up PyICE lint reports
- Fixed code tooltips for PyAI and PyICE
- Unified CodeTooltip (for PyAI and PyICE, PyLO and PyTRG still need to be unified)

poiuyqwert committed:
- Fixed up PyLO lint reports

poiuyqwert committed:
- Fixed up PyMPQ lint reports

poiuyqwert committed:
- Fixed up PyPAL lint reports

poiuyqwert committed:
- Fixed up PyPCX lint reports

poiuyqwert committed:
- Fixed up PySPK lint reports

poiuyqwert committed:
- Fixed up PyTBL lint reports

poiuyqwert committed:
- Fixed up PyTILE lint reports
- Fixed issue in PyTILE with copy/paste megatiles

poiuyqwert committed:
- Fixed up PyTRG lint reports

poiuyqwert committed:
- Fix layout issues with FixIssuesDialog

poiuyqwert committed:
- Added code preview to FixIssues dialog

poiuyqwert committed:
- Fix issue with decompiling an "empty" idle order flags param
- Added directives to expand unit/upgrade/tech id's

poiuyqwert committed:
- Move parser variable resolution into CodeType (this allows more complex types like UnitGroupCodeType to handle more cases)
- Added new directives to PyAI for manually managing expanded units/upgrades/tech
- Move some data to a new ParseSettings to allow it to flow through multiple ParseContext uses
- Fix UnitGroupCodeType and using variables in group
- Fix IdleOrderFlags parse/serialize with only empty BasicFlags

poiuyqwert committed:
Merge branch 'AISE' into strict-lint

# Conflicts:
#	PyMS/FileFormats/AIBIN/CodeHandlers/AISEIdleOrder.py
#	PyMS/FileFormats/AIBIN/CodeHandlers/CodeDirectives.py
#	PyMS/PyAI/CodeEditDialog.py
#	PyMS/PyAI/FixIssuesUI/FixIssuesDialog.py
#	PyMS/PyAI/PyAI.py

poiuyqwert committed:
- Fixed up tests lint reports

poiuyqwert committed:
- Allow raw flags for IssueOrderFlagsCodeType

poiuyqwert committed:
Merge branch 'AISE' into strict-lint

poiuyqwert committed:
- Fix AISEIdleOrder option types handling

poiuyqwert committed:
- Fixed up Utilities.CodeHandlers lint reports

poiuyqwert committed:
- Fixed up Utilities.SettingsUI lint reports

poiuyqwert committed:
- Fixed up Utilities.UIKit lint reports
- Fixed pollution from `import *` by adding `__all__` definitions to relavent files
- Remove some uses of `Any` where not strictly needed
- Remove all cases of inheriting from `object`

poiuyqwert committed:
- Fixed up CHK lint reports (except for implementing interpret functions)

poiuyqwert committed:
- Fixed up Utilities lint reports

poiuyqwert committed:
- Fixed up PyJSON lint reports (except unused index)

poiuyqwert committed:
- Fix issue with some optional dict params falling back to defaults when explicitly passing an empty dict
- Fix tooltips in PyTRG
- Fix issue with old load_file using encoding for binary files
- Fix ReportList weirdness with subclassing which caused PyMPQ to crash

poiuyqwert committed:
- Additional type hinting

poiuyqwert committed:
- Additional typing cleanup

poiuyqwert committed:
- Fix f-strings for pre 3.12 python versions

neivv commented on PyMS/FileFormats/AIBIN/CodeHandlers/AISEIdleOrder.py:
```diff
@@ -0,0 +1,904 @@
+				else:
+					existing_options.append(option)
+		if len(options) == 1 and OptionSet.TYPE_ID in options and len(options[OptionSet.TYPE_ID]) == 1:
+			self.options = cast(OptionSet, options[OptionSet.TYPE_ID][0]).options
```
This flattening is not valid, for example it is converting `Self(SpellEffects(Ensnare))` to `SpellEffects(Ensnare)` when no other flags are specified.

poiuyqwert replied:
Oh interesting, maybe I don't understand what `Self(...)` actually does. I would have assumed that those two examples result in the same thing. I guess I should just remove the flattening entirely?

BoneHorror replied:
Self() makes it so that the conditions are checked against the caster, not the target

poiuyqwert replied:
Ohhh, ok, that makes more sense. Thanks, I'll update this soon.

poiuyqwert replied:
This should be fixed now with the [latest commit](https://github.com/poiuyqwert/PyMS/pull/149/commits/cb62680193a1eb6b7a9c558f9c186303987f9015), and also makes the code a little more clean/obvious with separating out the `Self` logic from `OptionSet`, so now there is `OptionSet` and `SelfOptions`.

poiuyqwert committed:
- Fix issue with `Self` IdleOrder flags handling

poiuyqwert committed:
- Cleanup

poiuyqwert committed:
- Fix issue with saving long AISE LongBlockCodeType when file is expanded (expanded flag was not set on AISE context)

poiuyqwert committed:
- Small fixes
- Updated deprecated trace usage for newer tkinter support

poiuyqwert committed:
- Fixed frameless windows on newer versions of macOS
- Fixed PyAI error when extdef doesn't exist
- Fixed a few incorrect trace_add callback signatures
